torch/_inductor/fx_passes/ddp_fusion.py

```diff
@@ -0,0 +1,577 @@
+    Generator,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Union,
+)
+
+import torch
+import torch.fx as fx
+from torch._dynamo.utils import counters
+from torch.fx.passes.shape_prop import _extract_tensor_metadata, TensorMetadata
+from torch.utils._pytree import tree_flatten, tree_map, tree_unflatten
+
+from ..fx_utils import get_fake_args_kwargs
+from ..virtualized import V
+
+aten = torch.ops.aten
+logger: logging.Logger = logging.getLogger("comm_fusion")
+
+
+def move_block_after(block: List[fx.Node], target_node: fx.Node) -> None:
+    for node in block:
+        target_node.append(node)
+        target_node = node
+
+
+def move_block_before(block: List[fx.Node], target_node: fx.Node) -> None:
+    for node in block:
+        target_node.prepend(node)
+        target_node = node
+
+
+def call_function(
+    graph: fx.Graph,
+    target: Union[str, Callable[..., Any]],
+    args: Optional[Tuple[fx.node.Argument, ...]] = None,
+    kwargs: Optional[Dict[str, fx.node.Argument]] = None,
+) -> fx.Node:
+    # We accept target as a str to avoid typing error as the type of
+    # a node.target is Union[str, Callable[..., Any]].
+    # This also allows us to avoid writing check for every call.
+    if isinstance(target, str):
+        raise RuntimeError(f"Call function should not get a str target {target=}")
+    node = graph.call_function(target, args, kwargs)
+    _, args, kwargs = get_fake_args_kwargs(node)
+    with V.fake_mode:
+        node.meta["val"] = target(*args, **kwargs)
+        # node.meta["val"] may be a container. So we use tree_map here
+        # to recursively extract the tensor metadata.
+        node.meta["tensor_meta"] = tree_map(
+            _extract_tensor_metadata, (node.meta["val"],)
+        )[0]
+    return node
+
+
+@dataclass(unsafe_hash=True)
+class CommBlock:
+    shape: Union[torch.Size, List[torch.Size]]
+    node_list: List[fx.Node]
+    inputs: List[fx.Node]
+    wait_nodes: List[fx.Node]
+    comm_node: fx.Node
+    outputs: Set[fx.Node]
+
+
+def get_comm_block(comm_node: fx.Node) -> Optional[CommBlock]:
+    """
+    Given a collective node (e.g., allreduce), find out all the nodes belong to
+    this communcation.
+
+    Args:
+        comm_node(fx.Node): The target communication/collective node.
+    Returns:
+        The CommBlock that encapsulates the related nodes (e.g., wait_node) of
+        the given comm_node.
+    """
+    node_list = []
+    wait_nodes = []
+    inputs, _ = tree_flatten((comm_node.args, comm_node.kwargs))
+    input_nodes = [inp for inp in inputs if isinstance(inp, fx.Node)]
+    wait_prefixes = "wait_tensor"
+    # If the users of the wait node are following items, we consinder them
+    # to be a part of the output.
+    intermediate_outputs = ("split", "reshape", "getitem", "detach", "alias")
+
+    # The MAX_WAIT_DISTANCE is 2 because there is a getitem between the
+    # collective op and wait if the collective op is a coalesced op.
+    MAX_WAIT_DISTANCE = 2
```

Comment:
It looks like you're not quite decided between 2 extremes. 

A) you know what the graph from DDP tracing will be, and DDP always puts a comm+wait grouped together.
B) you don't know (or don't want to assume) what the graph from DDP tracing will be, so you implement a search function to find comm ... wait pairs

This code seems more like B but with this bound of MAX_WAIT_DISTANCE, which if incorrect, would lead to not finding (all of) the wait nodes.

Reply:
Is there a third option which is to check which wait node after this allreduce node in the graph is taking this allreduce's input as wait input? It might be more robust than setting `MAX_WAIT_DISTANCE`.

Or if setting `MAX_WAIT_DISTANCE` is preferred, wonder if we have tested that if `MAX_WAIT_DISTANCE` is too small in the future, one of the unit tests will fail?